lib/Db/LocalMailboxMessageMapper.php

```diff
@@ -0,0 +1,127 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCA\Mail\Db;
+
+use OCP\DB\Exception as DBException;
+use function array_map;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Db\QBMapper;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\IDBConnection;
+
+/**
+ * @template-extends QBMapper<LocalMailboxMessage>
+ */
+class LocalMailboxMessageMapper extends QBMapper {
+	/** @var LocalAttachmentMapper */
+	private $attachmentMapper;
+
+	/** @var RecipientMapper */
+	private $recipientMapper;
+
+	public function __construct(IDBConnection $db,
+								LocalAttachmentMapper $attachmentMapper,
+								RecipientMapper $recipientMapper) {
+		parent::__construct($db, 'mail_local_mailbox');
+		$this->recipientMapper = $recipientMapper;
+		$this->attachmentMapper = $attachmentMapper;
+	}
+
+	/**
+	 * @param string $userId
+	 * @return LocalMailboxMessage[]
+	 * @throws DBException
+	 */
+	public function getAllForUser(string $userId): array {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('m.*')
+			->from('mail_accounts', 'a')
+			->join('a', $this->getTableName(), 'm', $qb->expr()->eq('m.account_id', 'a.id'))
+			->where(
+				$qb->expr()->eq('a.user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR), IQueryBuilder::PARAM_STR),
+				$qb->expr()->eq('m.type', $qb->createNamedParameter(LocalMailboxMessage::OUTGOING, IQueryBuilder::PARAM_INT), IQueryBuilder::PARAM_INT)
+			);
+		$rows = $qb->execute();
+		$result = $rows->fetchAll();
+		$rows->closeCursor();
+		return $result;
+	}
+
+	/**
+	 * @throws DoesNotExistException
+	 */
+	public function find(int $id): LocalMailboxMessage {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->in('id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT), IQueryBuilder::PARAM_INT)
+			);
+		return $this->findEntity($qb);
+	}
+
+	public function getRelatedData(int $id, string $userId): array {
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('a.*')
+			->from('mail_lcl_mbx_attchmts', 'm')
+			->join('m', 'mail_attachments', 'a', $qb->expr()->eq('m.attachment_id', 'a.id'))
+			->where(
+				$qb->expr()->eq('a.user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR), IQueryBuilder::PARAM_STR),
+				$qb->expr()->eq('m.local_message_id', $qb->createNamedParameter($id, IQueryBuilder::PARAM_INT), IQueryBuilder::PARAM_INT)
+			);
+		$rows = $qb->execute();
+		$result = $rows->fetchAll();
+		$rows->closeCursor();
+		$related = [];
+		$related['attachments'] = $result;
+		$related['recipients'] = $this->recipientMapper->findRecipients($id, Recipient::MAILBOX_TYPE_OUTBOX);
```

Comment:
let's avoid these stringly typed arrays and attache the data to the LocalMailbox instances instead